File: 02-RockPaperScissorLizardSpock/rust/src/movement.rs
use std::vec::Vec;
use game_result::GameResult;

#[derive(PartialEq)]
pub enum Movement {
    ROCK,
    PAPER,
    SCISSORS,
    LIZARD,
    SPOCK
}

impl Movement {
    fn get_opponent_movements_to_win(&self) -> Vec<Movement> {
        match *self {
            Movement::ROCK => vec![Movement::SCISSORS, Movement::LIZARD],
            Movement::PAPER => vec![Movement::SPOCK, Movement::ROCK],
            Movement::SCISSORS => vec![Movement::LIZARD, Movement::PAPER],
            Movement::LIZARD => vec![Movement::PAPER, Movement::SPOCK],
            Movement::SPOCK => vec![Movement::ROCK, Movement::SCISSORS]
        }
    }

    pub fn compute_result(&self, opponent_movement: &Movement) -> GameResult {
        if self == opponent_movement {
            return GameResult::DRAW;
        }
        if self.get_opponent_movements_to_win().contains(opponent_movement) {
            return GameResult::PlayerOneWins;
        }
        return GameResult::PlayerTwoWins;
    }
}

File: 02-RockPaperScissorLizardSpock/kotlin/fbougeno/src/main/kotlin/com/ccsw/coedevon/codingdojo/rockpaperscissorslizardspock/Movement.kt
package com.ccsw.coedevon.codingdojo.rockpaperscissorslizardspock


enum class Movement {

    ROCK {
        override fun getOtherPlayerMovementsToWin(): Array<Movement> = arrayOf(Movement.SCISSORS, Movement.LIZARD)
    },
    PAPER {
        override fun getOtherPlayerMovementsToWin(): Array<Movement> = arrayOf(Movement.SPOCK, Movement.ROCK)
    },
    SCISSORS {
        override fun getOtherPlayerMovementsToWin(): Array<Movement> = arrayOf(Movement.LIZARD, Movement.PAPER)
    },
    LIZARD {
        override fun getOtherPlayerMovementsToWin(): Array<Movement> = arrayOf(Movement.PAPER, Movement.SPOCK)
    },
    SPOCK {
        override fun getOtherPlayerMovementsToWin(): Array<Movement> = arrayOf(Movement.ROCK, Movement.SCISSORS)
    };

    abstract fun getOtherPlayerMovementsToWin(): Array<Movement>

    fun calculateResult(otherMovement: Movement): GameResult {
        if (this == otherMovement)
            return GameResult.DRAW

        if (getOtherPlayerMovementsToWin().contains(otherMovement))
            return GameResult.PLAYER_1_WINS

        return GameResult.PLAYER_2_WINS
    }
}
